### ISSUE TITLE
"S&box lacks documentation" - Collection and discussion

### ISSUE THREAD

**Issue by yuberee** [Facepunch/sbox-issues#6850]
### Location

n/a

### Summary

The term "Documentation" is usually a scapegoat for why developers have no interest in S&box, in this issue I'll try to compile what's missing.
### Feel free to post your gripes here if you have any.

# Courses
Every major engine has _official_ courses, ranging from complete beginner to advanced:

- Roblox has [create.roblox.com](https://create.roblox.com/docs/tutorials);
- Unity has [learn.unity.com](https://learn.unity.com/courses);
- Unreal Engine has [academy.unrealengine.com](https://academy.unrealengine.com/totara/catalog/index.php);

The upside of having written courses compared to video tutorials is that they're easier to update in case there's any changes.
Currently docs.facepunch.com is good at explaining what systems do and how to use them, but useless in the eyes of a complete beginner.
Probably these courses will also need to explain the basics of C#, any extra work the user has to do to learn is developers lost.

# Api References aren't documented
https://sbox.game/api is full of broken or straight up missing references and descriptions.
Most methods, classes, and properties have XML summaries and remarks, they should automatically show up on the api pages as well.
Here's just a few examples of the most used classes:

- [Rotation.RotateAroundAxis](https://sbox.game/api/Rotation/RotateAroundAxis) - Completely missing summary and remarks;
- [TimeUntil](https://sbox.game/api/Sandbox.TimeUntil) - Summary isn't formatted correctly, missing references and new lines;
- [Search: Log.Info](https://sbox.game/api/where/is/Log.Info) - Global namespaces are missing;
- [Search: ClosestPoint](https://sbox.game/api/where/is/ClosestPoint) vs [Search: BBox.ClosestPoint](https://sbox.game/api/where/is/BBox.ClosestPoint) - More specific searches sometimes don't yield any results;

And generally it wouldn't hurt to have code snippetsor manual descriptions on the most used methods or the ones that get grossly unused (Still seems like most people don't understand how BBox.Rotate works)

And while we're at it, would be nice if you could click on methods and classes inside of the `code` sections in docs.facepunch.com, especially if courses get implemented.

# Still no templates
Templates were implemented **over 2 years ago** in June 2022, there are still no game templates available. I believe there are actually **less** compared to when the feature first launched.

I remember garry saying that the best tool for someone new to learn was addons, which was what made Gmod so easy to pick up. Modify a weapon stats, make them shoot exploding barrels, have the barrels roll towards players, and that could be a game by itself!

Not only are these templates great for piecing games together for anyone not confident enough to do it from scratch, they also serve the purpose of teaching correct programming conventions.

### Original templates:
![Image](https://github.com/user-attachments/assets/74a44c53-c3a1-4f96-97a5-94ce98c198f6)
### Current template(s?):
![Image](https://github.com/user-attachments/assets/597a279b-7e96-4093-ad8c-e325771c0d83)

# Regarding experience level
I'll bring up some points that were talked about in council chat, but I believe they're the general sentiment in the community.

When discussing what good Documentation criticism was, these were the kind of examples given:
- **Bad criticism**: There are no code examples;
- **Good criticism**: The description for **X** is misleading, it should say this instead...;

In what world does the "good criticism" come from the mouth of a complete beginner? And how is the "bad criticism" not valid?

In general, it seems like the issue is being looked at from the eyes of someone that, to be frank, doesn't need documentation to the level of what the target audience actually should be.
When people are complaining about documentation issues, they do so out of necessity, they're not doing it just to complain. (Althought I do agree most of the time it's used as a scapegoat from people too lazy to learn that feel called out)

Skilled programmers know how to navigate the api, know what to look for, how to look for it, decompile code, or find examples on github. The "documentation" they are looking for is fundamentally different from what the average Joe is asking for.
Statistically speaking we have to assume that the biggest audience is made up of complete beginners with a computer that is barely able to run S&box tools and Notepad++.

Like mentioned before feel free to discuss, I'm hoping I can shed some light on the point of view of S&box beginners but unfortunately I'm becoming jaded too.

**Comment by MrBrax** [comment]
Here are some I wrote up today, was about to make an issue after I would make some more:

- TypeLibrary usage and what methods can be used vs not whitelisted in Type
- Editor panels, control widgets, inputs, and how you use and design them in custom editors and editor apps
- Razor default elements and attributes like img, button, textentry
- Textures & Render target formats and what they're used for
- Material overrides, runtime material creation and attributes
- UI panel drag and drop, considering there are methods for it
- Gizmo usage, custom transforms
- Binding values in razor
- Razor render clarification, how to properly use buildhash and how child components render separately 
- Creating assets programatically in editor, and how to properly use absolute/relative paths when you don't have the FileSystem.Mounted context

**Comment by yuberee** [comment]
That's a lot of guides, they don't sound very beginner friendly though.

**Comment by matekdev** [comment]
Yeah, I'm honestly surprised to see templates haven't been brought back yet. It seems like a pretty essential feature for newcomers.

**Comment by handsomematt** [comment]
> That's a lot of guides, they don't sound very beginner friendly though.

Braxen is describing documentation, aren't you the one asking for guides

**Comment by garrynewman** [comment]
Fwiw I don't really agree that we should have templates just yet.

**Comment by yuberee** [comment]
With how volatile the API has been until now it would've been extra work for sure.
Just something to consider before doing any future push for developers, I'm sure the community would also be available to help, the wiki is not the best place for crowdsourcing.

**Comment by DrakeFruit** [comment]
community templates would be #6544

**Comment by Cauzzmo** [comment]
My only problem us the UI. Never been naturally good at UI but the UI system in the editor for s&box sucks so much fucking balls

**Comment by HaydenPrice** [comment]
I believe that a better Doc/API search would go a long way to alleviating many of the 'the documentation is bad' complaints. Having to know the exact word to search for when it comes to searching really hurts the limited documentation we do have.

<img width="1205" height="425" alt="Image" src="https://github.com/user-attachments/assets/c9c267e2-b79e-42c7-87ec-9ee8ae5ae2f9" />

vs.

<img width="1292" height="498" alt="Image" src="https://github.com/user-attachments/assets/fb4f3876-139d-42e9-ba98-cbbdaca6cb29" />

This also goes the other way, when you're searching for something broad - like a model - you get bombarded by irrelevance. 264 results for 'Model' and [Model](https://sbox.game/api/Sandbox.Model) is half way down the page - I have to scroll down to find it.

Having some kind of contextual or relevance search/results would add a lot of value.

<img width="1299" height="935" alt="Image" src="https://github.com/user-attachments/assets/643df5f2-ddd7-4359-bd97-85f8cf5ac345" />

I find a lot of the time if a search comes up with nothing, I have to spend 15 minutes digging through the tabs and looking through the sub-tabs to find what I am looking for. Take the _Execution Order_ example above. If I didn't know how bad (or specific) the search function was, I would have to start digging through tabs of documentation. (That is of course if I thought the documentation existed at all - it's not unreasonable to assume the documentation doesn't exist when you get 0 results for being 'pretty close')

So my adventure through the documentation begins;

- About -> Getting Started?
- Code -> Code basics?
- Code -> Advanced Topics?
- Scene -> Scenes?
- Scene -> GameObject?
- Scene -> Components?

Oh, I've finally managed to find Execution Order after an extra 5 minutes of having to skim documentation, because I didn't know that I had to search for _Execution_ instead of _Execute_.

Missing documentation and oddly categorized documentation is of course a problem that will be fixed/filled out over time, but I think the biggest immediate pain point is the way the searching works (or in many cases doesn't work). In the case of Execution Order isn't even oddly categorized - it's just relevant to a lot of things you just don't know which category it would be in.

You could have the most perfect formatting and categorization of the documentation/API - but still be held back by bad search results. But if you have a great - or even just good - search, it can do so much heavy lifting. I really doubt I'd care that much about how the documentation is categorized or formatted if I could have better and more relevant search results.

**Comment by MrBrax** [comment]
@HaydenPrice i talked about this recently, garry disagreed https://discord.com/channels/833983068468936704/833983416390385685/1387429532578873357

**Comment by HaydenPrice** [comment]
Adding more on to this today regarding a 'better search'.

I want to try and find the API page for widgets so I can see their base functions. Searching for widget provides exactly 0 links to the widget class itself.

<img width="1549" height="936" alt="Image" src="https://github.com/user-attachments/assets/d6bf4bf6-9af2-4e1c-a5f5-09c5dd63e28e" />

<img width="1816" height="945" alt="Image" src="https://github.com/user-attachments/assets/3d4e739e-a3b8-4fdc-ad25-c67d9a745985" />

Would be super helpful to a lot of people if [Editor.Widget](https://sbox.game/api/Editor.Widget) was first on the list, or at very least _on_ the list.

**Comment by handsomematt** [comment]
Make seperate issues
